After changing node ip and keeping its host_id, the old endpoint still appears in system.peers

Seen consistently in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/84/testReport/update_cluster_layout_tests/TestUpdateClusterLayout/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split012___test_change_node_ip/
For example:
```
    def check_peer_and_local_table(self, nodes):
        peers_and_local = []
        for node in nodes:
            status = nodetool_status(node)
            logger.debug("nodetool status from {}: {}".format(node.name, status))
            s = self.patient_exclusive_cql_connection(node)
            peers = rows_to_list(s.execute("SELECT host_id, peer FROM system.peers"))
            local = rows_to_list(s.execute("SELECT host_id, broadcast_address FROM system.local"))
>           peers_and_local.append(str(sorted(peers + local)))
E           TypeError: '<' not supported between instances of 'NoneType' and 'UUID'
```

Locally, I see also:
```
12:42:06,474 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2123 | Check peer table for node3 with ip address 127.0.49.33 : peers=[[UUID('fd396b2b-3162-4dc4-832d-fbe5777a2f14'), '127.0.49.2'], [UUID('eac15ed3-8f32-421b-91da-5d5bf5a0caa8'), '127.0.49.3'], [UUID('444777d5-1b55-4aa8-afd4-9d6b5f237fbd'), '127.0.49.1']] 
12:42:06,474 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2124 | Check peer table for node3 with ip address 127.0.49.33 : local=[[UUID('eac15ed3-8f32-421b-91da-5d5bf5a0caa8'), '127.0.49.33']] 
12:42:06,474 update_cluster_layout_tests WARNING update_cluster_layout_tests.py:2126 | Expected 2 peers but got 3
```

I'm not sure why this is seen consistently in debug mode and not in dev/release modes.
It's probably related to relative timing of racing async calls.

My analysis shows that `storage_service::on_join` is called with the old endpoint inet_address,
calling `on_change` with `application_state::STATUS` and `versioned_value::STATUS_NORMAL`,
calling `handle_state_normal(endpoint)`.

There log messages show:
```
WARN  2022-11-08 12:55:15,208 [shard 0] storage_service - Not updating host ID 81909e9f-202d-47ce-94da-3406dab51f26 for 127.0.63.3 because it's mine
```

BUT https://github.com/scylladb/scylladb/blob/a9063f9582f0ea22dd2c95d457c5626f23f0256e/service/storage_service.cc#L971-L977

calls `update_topology(endpoint, get_dc_rack_for(endpoint))`

since 421557b40aa8d1adf5016722418936a99595cb20

which puts it back into the topology.

And, https://github.com/scylladb/scylladb/blob/a9063f9582f0ea22dd2c95d457c5626f23f0256e/service/storage_service.cc#L997-L999
calls `notify_joined(endpoint)` which is not the root cause of this issue, but there's no reason to notify that the old endpoint joined when it owns no tokens and is marked for removal.

Later on, https://github.com/scylladb/scylladb/blob/a9063f9582f0ea22dd2c95d457c5626f23f0256e/service/storage_service.cc#L1217-L1218
Sees the old endpoint in token_metadata (after update_topology inserted it back)
and calls `do_update_system_peers_table(endpoint, state, value)` which adds it back to system.peers, after it was removed from it.

The following change fixes the issue for me locally:
```
diff --git a/service/storage_service.cc b/service/storage_service.cc
index 03611a7ed0..9896d4e6da 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -969,11 +969,15 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     }
 
     bool is_member = tmptr->is_member(endpoint);
+    bool do_notify_joined = !is_member && !endpoints_to_remove.contains(endpoint);
+
     // Update pending ranges after update of normal tokens immediately to avoid
     // a race where natural endpoint was updated to contain node A, but A was
     // not yet removed from pending endpoints
-    if (!is_member) {
-        tmptr->update_topology(endpoint, get_dc_rack_for(endpoint));
+    if (do_notify_joined) {
+        auto dc_rack = get_dc_rack_for(endpoint);
+        slogger.debug("handle_state_normal: update_topology: endpoint={} dc={} rack={}", endpoint, dc_rack.dc, dc_rack.rack);
+        tmptr->update_topology(endpoint, std::move(dc_rack));
     }
     co_await tmptr->update_normal_tokens(owned_tokens, endpoint);
     co_await update_pending_ranges(tmptr, format("handle_state_normal {}", endpoint));
@@ -994,7 +998,7 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     }
 
     // Send joined notification only when this node was not a member prior to this
-    if (!is_member) {
+    if (do_notify_joined) {
         co_await notify_joined(endpoint);
     }
 
```

--- a/db/system_keyspace.cc
+++ b/db/system_keyspace.cc
@@ -1502,6 +1502,7 @@ future<> system_keyspace::update_tokens(gms::inet_address ep, const std::unorder
     }
 
     sstring req = format("INSERT INTO system.{} (peer, tokens) VALUES (?, ?)", PEERS);
+    slogger.debug("INSERT INTO system.{} (peer, tokens) VALUES ({}, {})", PEERS, ep, tokens);
     auto set_type = set_type_impl::get_instance(utf8_type, true);
     co_await execute_cql(req, ep.addr(), make_set_value(set_type, prepare_tokens(tokens))).discard_result();
     co_await force_blocking_flush(PEERS);
@@ -1594,6 +1595,7 @@ future<> system_keyspace::update_peer_info(gms::inet_address ep, sstring column_
 
     co_await update_cached_values(ep, column_name, value);
     sstring req = format("INSERT INTO system.{} (peer, {}) VALUES (?, ?)", PEERS, column_name);
+    slogger.debug("INSERT INTO system.{} (peer, {}) VALUES ({}, {})", PEERS, column_name, ep, value);
     co_await execute_cql(req, ep.addr(), value).discard_result();
 }
 // sets are not needed, since tokens are updated by another method
@@ -1645,6 +1647,7 @@ future<> system_keyspace::update_schema_version(table_schema_version version) {
  */
 future<> system_keyspace::remove_endpoint(gms::inet_address ep) {
     sstring req = format("DELETE FROM system.{} WHERE peer = ?", PEERS);
+    slogger.debug("DELETE FROM system.{} WHERE peer = {}", PEERS, ep);
     co_await execute_cql(req, ep.addr()).discard_result();
     co_await force_blocking_flush(PEERS);
 }

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -984,7 +984,9 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
         // a race where natural endpoint was updated to contain node A, but A was
         // not yet removed from pending endpoints
         if (!is_member) {
-            tmptr->update_topology(endpoint, get_dc_rack_for(endpoint));
+            auto dc_rack = get_dc_rack_for(endpoint);
+            slogger.debug("handle_state_normal: update_topology: endpoint={} dc={} rack={}", endpoint, dc_rack.dc, dc_rack.rack);
+            tmptr->update_topology(endpoint, std::move(dc_rack));
             do_notify_joined = true;
         }
         co_await tmptr->update_normal_tokens(owned_tokens, endpoint);
@@ -997,7 +999,7 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     for (auto ep : endpoints_to_remove) {
         co_await remove_endpoint(ep);
     }
-    slogger.debug("handle_state_normal: endpoint={} owned_tokens = {}", endpoint, owned_tokens);
+    slogger.debug("handle_state_normal: endpoint={} is_member={} endpoint_to_remove={} owned_tokens={}", endpoint, is_member, endpoints_to_remove.contains(endpoint), owned_tokens);
     if (!owned_tokens.empty() && !endpoints_to_remove.count(endpoint)) {
         co_await update_peer_info(endpoint);
         try {
@@ -1229,6 +1231,7 @@ future<> storage_service::on_change(inet_address endpoint, application_state sta
             co_return;
         }
         if (get_token_metadata().is_member(endpoint)) {
+            slogger.debug("endpoint={} on_change:     updating system.peers table", endpoint);
             co_await do_update_system_peers_table(endpoint, state, value);
             if (state == application_state::RPC_READY) {
                 slogger.debug("Got application_state::RPC_READY for node {}, is_cql_ready={}", endpoint, ep_state->is_cql_ready());
@@ -1326,6 +1329,7 @@ future<> storage_service::do_update_system_peers_table(gms::inet_address endpoin
 }
 
 future<> storage_service::update_peer_info(gms::inet_address endpoint) {
+    slogger.debug("Update peer info: endpoint={}", endpoint);
     using namespace gms;
     auto* ep_state = _gossiper.get_endpoint_state_for_endpoint_ptr(endpoint);
     if (!ep_state) {

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -969,13 +969,27 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     }
 
     bool is_member = tmptr->is_member(endpoint);
-    // Update pending ranges after update of normal tokens immediately to avoid
-    // a race where natural endpoint was updated to contain node A, but A was
-    // not yet removed from pending endpoints
-    if (!is_member) {
-        tmptr->update_topology(endpoint, get_dc_rack_for(endpoint));
+    bool do_notify_joined = false;
+
+    if (endpoints_to_remove.contains(endpoint)) [[unlikely]] {
+        if (!owned_tokens.empty()) {
+            on_fatal_internal_error(slogger, format("endpoint={} is marked for removal but still owns {} tokens", endpoint, owned_tokens.size()));
+        }
+    } else {
+        if (owned_tokens.empty()) {
+            on_internal_error_noexcept(slogger, format("endpoint={} is not marked for removal but owns no tokens", endpoint));
+        }
+
+        // Update pending ranges after update of normal tokens immediately to avoid
+        // a race where natural endpoint was updated to contain node A, but A was
+        // not yet removed from pending endpoints
+        if (!is_member) {
+            tmptr->update_topology(endpoint, get_dc_rack_for(endpoint));
+            do_notify_joined = true;
+        }
+        co_await tmptr->update_normal_tokens(owned_tokens, endpoint);
     }
-    co_await tmptr->update_normal_tokens(owned_tokens, endpoint);
+
     co_await update_pending_ranges(tmptr, format("handle_state_normal {}", endpoint));
     co_await replicate_to_all_cores(std::move(tmptr));
     tmlock.reset();
@@ -994,7 +1008,7 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     }
 
     // Send joined notification only when this node was not a member prior to this
-    if (!is_member) {
+    if (do_notify_joined) {
         co_await notify_joined(endpoint);
     }
 

